pkg/clustermesh/kvstoremesh/kvstoremesh.go

```diff
@@ -105,5 +167,32 @@ func (km *KVStoreMesh) newRemoteCluster(name string, _ common.StatusFunc) common
 	run(rc.identities.syncer.Run)
 	run(rc.ipcache.syncer.Run)
 
+	run(rc.waitForConnection)
+
 	return rc
 }
+
+// synced returns once all remote clusters have been synchronized or the global
+// timeout has been reached. The given syncCallback is always executed before
+// the function returns.
+func (km *KVStoreMesh) synced(ctx context.Context, syncCallback func(context.Context)) error {
+	ctx, cancel := context.WithTimeout(ctx, km.config.GlobalReadyTimeout)
+	defer func() {
+		syncCallback(ctx)
+		cancel()
+	}()
+
+	waiters := make([]wait.Fn, 0)
+	km.common.ForEachRemoteCluster(func(rci common.RemoteCluster) error {
+		rc := rci.(*remoteCluster)
+		waiters = append(waiters, rc.synced.Resources)
+		return nil
+	})
+
+	if err := wait.ForAll(ctx, waiters); err != nil {
+		km.logger.WithError(err).Info("Failed to wait for synchronization. KVStoreMesh will now handle requests, but some clusters may not have been synchronized.")
```

Comment:
nit: Probably could be a warning and also it would be nice to provide information for which cluster it happened for easier debugging (Doesn't need to be a list of all clusters, just an example will be helpful I think)

Reply:
Made this Info level because it seemed similar in vein to the connection timeout, since it doesn't necessarily mean the synchronization itself failed, just that it is taking longer than expected: https://github.com/cilium/cilium/pull/30361#discussion_r1489590058

We could add this info to the log, but the data is also available via the `kvstore_initial_sync_completed` metric that kvstoremesh exports